text/3691-trusted-publishing-cratesio.md

```diff
@@ -0,0 +1,371 @@
+9. As a good steward, once the crate has successfully been published, the workflow job will request the revocation of the access token from crates.io
+    - An access token's lifetime might outlive the actual time needed to perform the action. It's good practice to revoke it once done to further limit any risks of accidental exposure. 
+
+> [!NOTE]
+> Steps 7 and 8 are abstracted in the `cargo publish` command. The access token would be passed via the `CARGO_REGISTRY_TOKEN` environment variable.
+ 
+</details>
+
+### Trusted Publisher Configuration on crates.io
+[trusted-publisher-configuration-cratesio]: #trusted-publisher-configuration-cratesio
+
+Prior to publishing a crate through a GitHub Actions workflow, the crate author will need to create a _Trusted Publisher Configuration_ via the crates.io web UI. This is a one-time configuration **required per crate**.
+
+> [!Note]
+> Initially, a _Trusted Publisher Configuration_ can only be created after an initial manual publishing of a crate.
+> 
+> See [Future Possibilities](#future-possibilities) for future plans to alleviate this pain.
+
+Without a corresponding _Trusted Publisher Configuration_ any attempts from a GitHub Action workflow to exchange an ID token for an access token from crates.io will be unauthorized and fail. 
+
+A _Trusted Publisher Configuration_ will vary on requirements depending on the trusted publisher being used (e.g. GitHub Actions, GitLab Pipelines, and so on). CI/CD providers will vary in the claims they make within their ID token. 
+
+For GitHub Actions, the crate author will need to provide:
+
+1. (required) The owning GitHub username or organization
+2. (required) The repository name
+3. (required) The workflow file name (must be located in `.github/workflows/`)
+4. (optional) The [GitHub Actions environment](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment) name
+
+
+### Example GitHub Actions Workflow
+[example-gha-workflow]: #example-gha-workflow
+
+With a _Trusted Publisher Configuration_ in place, a example GitHub Actions workflow may look like this:
+
+```yaml
+name: Publish Crate
+
+# Some may opt to use a trigger based on a push with a tag instead
+on:
+  release:
+    types: [published]
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    
+    # here we use a GitHub Actions _Environment_ that allows for more 
+    # control and steps to ensure security of the publishing 
+    # process
+    # 
+    # See https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment
+    environment: release
+
+    permissions:
+      id-token: write  # required in order to get a signed ID token
+      contents: read
+
+    steps:
+    - name: Checkout code
+      uses: actions/checkout@v4
+
+    - name: Install dependencies
+      run: cargo build --release
```

Comment:
This makes sure the crate successfully builds before acquiring the token. If the crate takes 1+ hour to build, the token might expire during the `cargo publish` verification checks.

Reply:
Note some [related discussion below](https://github.com/rust-lang/rfcs/pull/3691#discussion_r1763748068), where one idea was `cargo publish` itself could manage token provisioning and do it immediately before uploading an already built crate to avoid this problem

Reply:
I added a comment in the workflow definition to indicate that this is not strictly necessary.

[Related to the discussion below](https://github.com/rust-lang/rfcs/pull/3691/files#r1817443580), I'd like to implement the token exchange support initially as a separate action instead of directly within `cargo`. This will keep the work de-coupled.